examples/dataflow-hashpipeline/hashpipeline/buffering.py

```diff
@@ -0,0 +1,143 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# pytype: skip-file
+
+from __future__ import absolute_import
+from apache_beam.transforms.userstate import BagStateSpec, TimerSpec, CombiningValueStateSpec, on_timer
+from apache_beam.transforms.core import CombineValues
+from apache_beam.transforms.timeutil import TimeDomain
+from apache_beam.coders import VarIntCoder, TupleCoder, StrUtf8Coder
+from apache_beam.transforms.combiners import CountCombineFn
+import apache_beam as beam
+import time
+
+class PayloadFactory:
+  def __init__(self, filename, max_size):
+    self.payloads = []
+    self.max_size = max_size
+    self.filename = filename
+
+  def running_payload(self):
+    '''Returns a singleton instance of the current payload'''
+    raise NotImplementedError
+  
+  def flatten(self):
+    '''Adds current payload to payloads list and flattens'''
+    self.payloads.append(self.running_payload())
+    return [(self.filename, payload) for payload in self.payloads]
+  
+  def extend_running_payload(self, data):
+    '''Extends the current_singleton object until extending it further would pass max_size'''
+    raise NotImplementedError
+
+class FilePayloadFactory(PayloadFactory):
+  '''Payload factory for buffering lines in a file before sending batch requests to DLP'''
+  MAX_BYTE_SIZE_PER_REQUEST = 400000 # 0.4MB giving 100KB for metadata
+
+  def __init__(self, filename, max_size = MAX_BYTE_SIZE_PER_REQUEST):
+    super(FilePayloadFactory, self).__init__(filename, max_size) 
+    self._current = ""
+  
+  def running_payload(self):
+    return self._current
+  
+  # NOTE: This does not handle the case where a single line is longer than max_size
+  def extend_running_payload(self, data):
+    if len(self._current) + len(data) > self.max_size:
+      self.payloads.append(self._current)
+      self._current = ""
+    else:
+      self._current += data
+
+class ListPayloadFactory(PayloadFactory):
+  '''Payload factory for buffering SSN hash digests to batch get from Firestore'''
+  MAX_KEYS_PER_REQUEST = 1000 # 10MB max request size
+
+  def __init__(self, filename, max_size = MAX_KEYS_PER_REQUEST):
+    super(ListPayloadFactory, self).__init__(filename, max_size)
+    self._current = []
+  
+  def running_payload(self):
+    return self._current
+  
+  # NOTE: This does not handle the case where a single line is longer than max_size
+  def extend_running_payload(self, data):
+    if len(self._current) + len(data) > self.max_size:
+      self.payloads.append(self._current)
+      self._current = []
+    else:
+      self._current.append(data)
+
+
+class StatefulBufferingFn(beam.DoFn):
+  MAX_BUFFER_SIZE = 500 # 500 lines
+  BUFFER_STATE = BagStateSpec('buffer', TupleCoder((StrUtf8Coder(), StrUtf8Coder())))
+  COUNT_STATE = CombiningValueStateSpec('count', VarIntCoder(), CountCombineFn())
+  EXPIRY_TIMER = TimerSpec('expiry', TimeDomain.WATERMARK)
```

Comment:
I don't think  expiry / allowed lateness make sense for this use case.
on what timestamp is window assignment / lateness evaluated against?